diff -ur pkgconf-0.9.12/main.c pkgconf-0.9.12bero/main.c
--- pkgconf-0.9.12/main.c	2015-07-19 23:33:16.000000000 +0200
+++ pkgconf-0.9.12bero/main.c	2016-08-24 01:11:00.921650365 +0200
@@ -182,6 +182,18 @@
 }
 
 static void
+print_provides(pkg_t *pkg, void *unused, unsigned int flags)
+{
+	(void) unused;
+	(void) flags;
+
+	printf("%s", pkg->realname);
+	if (pkg->version != NULL)
+		printf(" = %s", pkg->version);
+	putchar('\n');
+}
+
+static void
 print_variables(pkg_t *pkg, void *unused, unsigned int flags)
 {
 	pkg_node_t *node;
@@ -268,6 +280,19 @@
 }
 
 static bool
+apply_printprovides(pkg_t *world, void *unused, int maxdepth, unsigned int flags)
+{
+	int eflag;
+
+	eflag = pkg_traverse(world, print_provides, unused, maxdepth, flags);
+
+	if (eflag != PKG_ERRF_OK)
+		return false;
+
+	return true;
+}
+
+static bool
 apply_modversion(pkg_t *world, void *unused, int maxdepth, unsigned int flags)
 {
 	int eflag;
@@ -902,10 +927,17 @@
 
 	if ((want_flags & PKG_PROVIDES) == PKG_PROVIDES)
 	{
-		fprintf(stderr, "pkgconf: warning: --print-provides requested which is intentionally unimplemented, along with\n");
+		fprintf(stderr, "pkgconf: warning: --print-provides requested which is intentionally implemented incompletely, along with\n");
 		fprintf(stderr, "  the rest of the Provides system.  The Provides system is broken by design and requires loading every\n");
 		fprintf(stderr, "  package module to calculate the dependency graph.  In practice, there are no consumers of this system\n");
 		fprintf(stderr, "  so it will remain unimplemented until such time that something actually uses it.\n");
+
+		want_flags &= (PKG_CFLAGS|PKG_LIBS);
+		if (!pkg_queue_apply(&pkgq, apply_printprovides, 2, global_traverse_flags, NULL))
+		{
+			ret = EXIT_FAILURE;
+			goto out;
+		}
 	}
 
 	if ((want_flags & PKG_DIGRAPH) == PKG_DIGRAPH)
